### ISSUE TITLE
About the Total Loss Function

### ISSUE THREAD

**Issue by xudh777** [abhi1kumar/DEVIANT#24]
Encountered a problem that I haven't quite understood. This problem exists in multiple 3D monocular object detection algorithms
The total loss in the algorithm is obtained by adding up multiple loss terms, However, adding the loss term of aa loss during the training process may result in negative values, as shown in the above figure. Will this simple summation method have a negative impact on the total loss??
![loss1](https://github.com/abhi1kumar/DEVIANT/assets/52651891/bb3de9cb-9a1c-405d-ae63-68628a6e2d4f)
The total loss in the algorithm is obtained by adding multiple loss terms, but I have found that some loss terms may be negative. I would like to ask if this negative loss value will have a negative impact on the total loss.
I found through research that negative losses come from laplacian_aleatoric_uncertainty_loss
![loss](https://github.com/abhi1kumar/DEVIANT/assets/52651891/2a124b18-3fdc-4134-808e-067fd747ee52)
I have also tried to increase the absolute value of the loss terms that result in negative numbers, so that all the loss terms are positive, but the resulting model performance is not ideal
I have not fully understood this position and hope to receive guidance. Looking forward to your reply. Thank you very much

**Comment by abhi1kumar** [comment]
Hi @xudh1991 
Thank you for putting this question in a separate issue from #23.

> adding the loss term of a loss during the training process may result in negative values, as shown in the above figure. Will this simple summation method have a negative impact on the total loss??

No, this does not impact the total loss. We encountered this issue in our [LUVLi facial landmark detection](https://openaccess.thecvf.com/content_CVPR_2020/papers/Kumar_LUVLi_Face_Alignment_Estimating_Landmarks_Location_Uncertainty_and_Visibility_Likelihood_CVPR_2020_paper.pdf) paper.

**Answer 1** 
Loss merely indicates disagreement of the prediction from the ground truth. However, neural networks use the **loss gradient** and not the loss directly during training. In other words, the gradients matter and not the loss values directly. This means you can shift the loss value to negative by adding a constant and still get the same gradient.

For example, consider the MSE loss function `0.5 * (p - y)^2`, where `p` is the prediction and `y` is the target ground truth. You can always define another loss function `0.5 * (p - y)^2 - 100`, which makes your loss function go negative. However, despite different loss values, the gradients in both cases are exactly equal.

**Answer 2**
A more technical answer involves probabilistic machine learning. Loss functions are the negative log-likelihoods of the probability density estimates, which means we can simplistically write:
``` Loss (x) = - log ( P(x) )```

The choice of different likelihoods or PDF `P(x)` leads to different loss functions.

P(x)   | Loss
----     | -------
Laplacian with fixed uncertainty | MAE
Laplacian with learnable uncertainty | Aleatoric MAE
Gaussian with fixed uncertainty | MSE
Gaussian with learnable uncertainty | Aleatoric MSE
[Mixed](https://openaccess.thecvf.com/content_CVPR_2020/papers/Kumar_LUVLi_Face_Alignment_Estimating_Landmarks_Location_Uncertainty_and_Visibility_Likelihood_CVPR_2020_paper.pdf) | Sum of classification + regression

Because `P(x)` is a PDF, sometimes it can be **greater** than 1. Taking the negative log-likelihood means you take the negative log of a quantity greater than 1, which is negative. Hence, the loss shows up negative.

The loss function or the negative log-likelihood of Laplacian or Gaussian random variables with variable uncertainty is the sum of two terms: a regularization term (log of covariance) and a Mahalanobis distance. See equations 11 and 13 of [LUVLi](https://openaccess.thecvf.com/content_CVPR_2020/papers/Kumar_LUVLi_Face_Alignment_Estimating_Landmarks_Location_Uncertainty_and_Visibility_Likelihood_CVPR_2020_paper.pdf) paper and ignore the first two classification terms on the RHS.

Returning to your question, the depth and size loss in DEVIANT / GUPNet  are probabilistic random variables. The regularization term is negative, while the Mahalanobis term is always positive. The regularization term is more negative than the Mahalanobis term. Therefore, the final depth and size loss that gets printed is negative.

**Comment by abhi1kumar** [comment]
Closing due to inactivity.

**Comment by xudh777** [comment]
Thank you very much. Although I don't quite understand, there seems to be some clarity
